delphi/app/graph.py

```diff
@@ -1,75 +1,280 @@
+        self.vertices.append(new_vertex)
+        self.network[new_vertex.name] = []
 
-        TODO: Have this do calculations on how to update the capacity rather than just setting it.
+    def add_edge(self, start, end, capacity):
         """
+        Adds a directed edge to the flow network
+        :param start: The node the edge is directed out of
+        :param end: The node the edge is directed into
+        :param capacity: The initial amount that can flow through the edge
+        :return: string if error occurs, nothing otherwise
+        """
+        if start == end:
+            return "Cannot have same start and end"
+        if not self.vertex_in_network(start):
+            return "Start vertex has not been added yet"
+        if not self.vertex_in_network(end):
+            return "End vertex has not been added yet"
+        for edge_check in self.get_edges():
+            if edge_check.start == start and edge_check.end == end and edge_check.capacity == capacity:
+                return "STOP"
+        new_edge = Edge(start, end, capacity)
+        return_edge = Edge(end, start, 0)
+        new_edge.returnEdge = return_edge
+        return_edge.returnEdge = new_edge
+        vertex = self.get_vertex(start)
+        self.network[vertex.name].append(new_edge)
+        return_vertex = self.get_vertex(end)
+        self.network[return_vertex.name].append(return_edge)
+
+    def get_path(self, start, end, path):
+        """
+        Finds a path from a start node to an end node
+        :param start: the node to start searching from
+        :param end: The node to arrive at
+        :param path: The set of nodes and edges to look for a path through
+        :return: the found path
+        """
+        if start == end:
+            return path
+        for edge_check in self.network[start]:
+            residual_capacity = edge_check.capacity - edge_check.flow
+            if residual_capacity > 0 and not (edge_check, residual_capacity) in path:
+                result = self.get_path(edge_check.end, end, path + [(edge_check, residual_capacity)])
+                if result is not None:
+                    return result
+
+    def calculate_max_flow(self):
+        """
+        Calculates the max flow of the flow network using the Ford-Fulkerson algorithm
+        :return: and integer representing the max flow
+        """
+        source = self.get_source()
+        sink = self.get_sink()
+        if source is None or sink is None:
+            return "Network does not have source and sink"
+        path = self.get_path(source.name, sink.name, [])
+        while path is not None:
+            flow = min(edge[1] for edge in path)
+            for edge, res in path:
+                edge.flow += flow
+                edge.returnEdge.flow -= flow
+            path = self.get_path(source.name, sink.name, [])
+        return sum(edge.flow for edge in self.network[source.name])
+
+
+def run_optimization(PKI_model):
```

Comment:
I would recommend making this part of the analyzer

Reply:
What would be the point in doing it in the analyzer as opposed to where it is now? Just seems like it would be alot of logic cluttering up the plugin

Reply:
The analyzer should compute the results of the data for whatever input and prior state it has. This function depends on the state of PKI/the flow network, so putting it in the analyzer keeps the functionality in the "proper" place and makes it easier to follow

Reply:
So I can see where you are coming from on this. I can go ahead and change it if you think its worth the time (because it will take a hefty bit of refactoring). Otherwise, we are running out of time and efforts could be put elsewhere.

Reply:
eh ill just go ahead and do it.